src/algebra/group/freiman.lean

```diff
@@ -0,0 +1,337 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.fun_like
+import data.multiset.basic
+
+/-!
+# Freiman homomorphisms
+
+In this file, we define Freiman homomorphisms. A `n`-Freiman homomorphism is a function `f : α → β`
+such that `f (x₁) * ... * f (xₙ) = f (y₁) * ... * f (yₙ)` for all `x₁, ..., xₙ, y₁, ..., yₙ` such
+that `x₁ * ... * xₙ = y₁ * ... * yₙ`. In particular, any `mul_hom` is a Freiman homomorphism.
+
+They are of interest in additive combinatorics.
+
+## Main declaration
+
+* `freiman_hom`: Freiman homomorphism.
+* `add_freiman_hom`: Additive Freiman homomorphism.
+
+## Notation
+
+* `α →*[n] β`: Multiplicative `n`-Freiman homomorphism
+* `α →+[n] β`: Additive`n`-Freiman homomorphism
+
+## TODO
+
+`monoid_hom.to_freiman_hom` could be relaxed to `mul_hom.to_freiman_hom` by proving
+`(s.map f).prod = (t.map f).prod` directly by induction instead of going through `f s.prod`.
+-/
+
+open multiset
+
+variables {F α β γ δ G : Type*}
+
+/-- An additive `n`-Freiman homomorphism is a map which preserves sums of `n` elements. -/
+structure add_freiman_hom (α β : Type*) [add_comm_monoid α] [add_comm_monoid β] (n : ℕ) :=
+(to_fun : α → β)
+(map_sum_eq_map_sum' {s t : multiset α} (hs : s.card = n) (ht : t.card = n) (h : s.sum = t.sum) :
+  (s.map to_fun).sum = (t.map to_fun).sum)
+
+/-- A `n`-Freiman homomorphism is a map which preserves products of `n` elements. -/
+@[to_additive add_freiman_hom]
+structure freiman_hom (α β : Type*) [comm_monoid α] [comm_monoid β] (n : ℕ) :=
```

Comment:
Thinking about this a little more, I'm not certain this is the convenient notion in the context of additive combinatorics. Following the syntax to the letter, it's usually stated as a map between subsets of abelian groups satisfying axioms, rather than a map between the whole groups themselves; of course the abelian group -> comm monoid generalisation is a sensible one, but the subset part worries me slightly. Since we have a coercion from `set α` to `Type*`, you could still be okay, but the subset might not be multiplicatively closed. More specifically I'm not certain that the diameter result in [Example 7.25 here and the opposite in Example 7.26](https://yufeizhao.com/gtac/fa17/gtac.pdf) can be nicely formalised because the subset `A` won't be closed under the group operation. (A special case of this is what gives us that finding 3APs in Z is the same thing as 3APs in Z/nZ for large enough n which is relevant for other proofs of Roth's theorem and related results). 

Perhaps an appropriate adjustment could be something like
```
structure freiman_hom_on (α β : Type*) [comm_monoid α] [comm_monoid β] (A : set α) (B : set β) (n : ℕ) :=
(to_fun : α → β)
(image_subset : A.image to_fun ⊆ B
(map_prod_eq_map_prod' {s t : multiset α} (hs : s.card = n) (ht : t.card = n)
  (h : s.prod = t.prod) {hs' : s ⊆ A} {ht' : t ⊆ B} : (s.map to_fun).prod = (t.map to_fun).prod)
```
maybe? Or the function might be only defined on `A`?